### PR TITLE
fix: calculate/add MD5 hash for multiple object delete.

### LINKED ISSUE #579: Tests fail with master Minio but not with play.minio.io
./gradlew runFunctionalTest gives the following error:

```
Test: removeObject(final String bucketName, final Iterable<String> objectNames)                                                                                  [11/644]
error occured
ErrorResponse(code=MissingContentMD5, message=Missing required header for this request: Content-Md5., bucketName=null, objectName=null, resource=/minio-java-test-1mdiula
, requestId=3L137, hostId=3L137)
request={method=POST, url=http://localhost:9001/minio-java-test-1mdiula?delete=, headers=Host: localhost:9001
User-Agent: Minio (amd64; amd64) minio-java/dev
x-amz-content-sha256: b542974fce9a1bea0370119bc29d681fbf85d9e9fddadb0bbad1cea20a1035e9
x-amz-date: 20170521T183721Z
Authorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20170521/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*
}
response={code=400, headers=Accept-Ranges: bytes
Content-Type: application/xml
Server: Minio/DEVELOPMENT.2017-05-03T09-04-56Z (linux; amd64)
Vary: Origin
X-Amz-Bucket-Region: us-east-1
X-Amz-Request-Id: 14C0B2F4F7E0175D
Date: Sun, 21 May 2017 18:37:21 GMT
Transfer-Encoding: chunked
}

        at io.minio.MinioClient.execute(MinioClient.java:1104)
        at io.minio.MinioClient.executePost(MinioClient.java:1251)
        at io.minio.MinioClient.removeObject(MinioClient.java:2176)
        at io.minio.MinioClient.access$000(MinioClient.java:158)
        at io.minio.MinioClient$4$1.populate(MinioClient.java:2254)
        at io.minio.MinioClient$4$1.hasNext(MinioClient.java:2275)
        at FunctionalTest.removeObject_test2(FunctionalTest.java:724)
        at FunctionalTest.runTests(FunctionalTest.java:1549)
        at FunctionalTest.main(FunctionalTest.java:1630)
:functional:runFunctionalTest FAILED
```

### PR DESCRIPTION
**Pull request by balamurugana**
Fixes #579

### CODE REVIEW

**Comment by balamurugana** [comment]
All functional tests pass with Amazon AWS S3

File: api/src/main/java/io/minio/Digest.java [hunk 1]
```diff
@@ -42,84 +43,85 @@ private Digest() {}
...
-  public static String md5Hash(String string) throws NoSuchAlgorithmException {
-    return md5Hash(string.getBytes(StandardCharsets.UTF_8));
-  }
 
+    if (data instanceof BufferedInputStream || data instanceof RandomAccessFile) {
+      updateDigests(data, len, sha256Digest, null);
+    } else if (data instanceof byte[]) {
+      sha256Digest.update((byte[]) data, 0, len);
+    } else {
+      throw new InternalException("unknown data to calculate SHA256 hash.  This should not happend.  "
```
Comment (harshavardhana) [inline api/src/main/java/io/minio/Digest.java]: `Unknown data source to calculate sha256 hash. This should not happen, please report this issue at https://github.com/minio/minio-java/issues` - is better i think.

File: api/src/main/java/io/minio/Digest.java [hunk 2]
```diff
@@ -42,84 +43,85 @@ private Digest() {}
...
 
-    messageDigest.update(data, 0, length);
+    if (data instanceof BufferedInputStream || data instanceof RandomAccessFile) {
+      updateDigests(data, len, sha256Digest, md5Digest);
+    } else if (data instanceof byte[]) {
+      sha256Digest.update((byte[]) data, 0, len);
+      md5Digest.update((byte[]) data, 0, len);
+    } else {
+      throw new InternalException("unknown data to calculate SHA256 hash.  This should not happend.  "
+                                  + "Please report this issue at https://github.com/minio/minio-java/issues");
```
Comment (harshavardhana) [inline api/src/main/java/io/minio/Digest.java]: Same as above..

File: api/src/main/java/io/minio/Digest.java [hunk 3]
```diff
@@ -42,84 +43,85 @@ private Digest() {}
...
+    throws NoSuchAlgorithmException, IOException, InsufficientDataException, InternalException {
     MessageDigest md5Digest = MessageDigest.getInstance("MD5");
-    updateDigests(inputStream, len, null, md5Digest);
+
+    if (data instanceof BufferedInputStream || data instanceof RandomAccessFile) {
+      updateDigests(data, len, null, md5Digest);
+    } else if (data instanceof byte[]) {
+      md5Digest.update((byte[]) data, 0, len);
+    } else {
+      throw new InternalException("unknown data to calculate SHA256 hash.  This should not happend.  "
```
Comment (harshavardhana) [inline api/src/main/java/io/minio/Digest.java]: Same as above..

File: api/src/main/java/io/minio/Digest.java [hunk 4]
```diff
@@ -42,84 +43,85 @@ private Digest() {}
...
+                                  + "Please report this issue at https://github.com/minio/minio-java/issues");
+    }
+
     return BaseEncoding.base64().encode(md5Digest.digest());
   }
 
+
```
Comment (harshavardhana) [inline api/src/main/java/io/minio/Digest.java]: Unexpected line?

File: api/src/main/java/io/minio/MinioClient.java [hunk 5]
```diff
@@ -896,40 +897,31 @@ public void writeTo(BufferedSink sink) throws IOException {
...
-        // Fix issue #567: Compute SHA256 hash only.
-        if (body == null) {
-          sha256Hash = Digest.sha256Hash(new byte[0]);
-        } else if (body instanceof BufferedInputStream) {
-          sha256Hash = Digest.sha256Hash((BufferedInputStream) body, length);
-        } else if (body instanceof RandomAccessFile) {
-          sha256Hash = Digest.sha256Hash((RandomAccessFile) body, length);
-        } else if (body instanceof byte[]) {
-          sha256Hash = Digest.sha256Hash((byte[]) body, length);
+        Object data = body;
```
Comment (poornas) [inline api/src/main/java/io/minio/MinioClient.java]: you might want to consider ternary operator for conciseness
Object data = (body == null) ? new byte[0] : body;
int len = (body == null) ? 0 : length;

File: api/src/main/java/io/minio/Digest.java (hunk 1 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/Digest.java]: done

File: api/src/main/java/io/minio/Digest.java (hunk 4 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/Digest.java]: no. you can see double newlines between methods

File: api/src/main/java/io/minio/MinioClient.java (hunk 5 above)
Comment (balamurugana) [inline api/src/main/java/io/minio/MinioClient.java]: For this case, `data` and `len` are reset when `data == null`  in one `if` seems better to me for readability sake.

There is a chance of misread two `(body == null)` statements.

**Pull request merged**